### PR TITLE
[fully_shard][DTensor] Support fully_shard with DTensor on full SPMD mesh

### PR DESCRIPTION
**Pull request by fegin**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.14.0) (oldest at bottom):
* __->__ #176334

**Summary**
Enable `fully_shard` to operate on models whose parameters are already DTensors distributed across a full SPMD mesh, including the data-parallel dimensions.

Previously, `fully_shard` only handled the case where DP dimensions were not part of an existing DTensor mesh (which is TP or EP). This PR introduces `DataParallelMeshDims`, which lets users specify which mesh dimensions correspond to data parallelism (sharding and/or replication). The DP dimensions must carry `Replicate` placements in the original DTensor spec before `fully_shard` replaces them with `Shard`.

Note that, the input/activation will also be DTensors distributed across the full SPMD mesh.

**Verification**
We verify this PR through unittests and TorchTitan integration: https://github.com/pytorch/torchtitan/pull/2616.

### CODE REVIEW

File: test/distributed/_composable/fsdp/test_fully_shard_dtensor.py [hunk 1]
```diff
@@ -0,0 +1,400 @@
...
+            inp = torch.randn((2, mlp_dim), device=device_type)
+            ref_optim.zero_grad(set_to_none=(i % 2 == 0))
+            ref_loss = ref_model(inp).sum()
+            ref_loss.backward()
+            ref_optim.step()
+
+            optim.zero_grad(set_to_none=(i % 2 == 0))
+            if mesh is not None:
+                inp = DTensor.from_local(
+                    inp, mesh, [Replicate()] * mesh.ndim, run_check=False
```
Comment (weifengpy) [inline test/distributed/_composable/fsdp/test_fully_shard_dtensor.py:80]: noob: why it's Replicate() instead of Shard(0) on the DP (batch) dim?

File: test/distributed/_composable/fsdp/test_fully_shard_dtensor.py [hunk 2]
```diff
@@ -0,0 +1,400 @@
...
+        mesh = init_device_mesh(
+            device_type.type, (self.world_size,), mesh_dim_names=("fsdp",)
+        )
+        dp_pg = mesh.get_group()
+
+        torch.manual_seed(42)
+        model = MLP(16, device=device_type)
+        ref_model = copy.deepcopy(model)
+
+        distribute_module(model, mesh)
```
Comment (weifengpy) [inline test/distributed/_composable/fsdp/test_fully_shard_dtensor.py]: my ask to claude "what does fully_shard do if user forgot to call distribute_module, but pass fully_shard global spmd mesh?"

```
1. Validation passes — _validate_mesh only checks that the named dims exist in the mesh. It doesn't check whether parameters are DTensors.                                                                                                
  2. mesh_info is constructed with is_spmd_mesh=True — because dp_mesh_dim_names was provided. The DP submesh is extracted from the full mesh.
  3. In FSDPParam.__init__ — self.is_dtensor = isinstance(param, DTensor) is False since params are plain tensors.                                                                                                                          
  4. Dispatch in _init_sharding_spec (line 311-315):                                                                                                                                                                                      
  if self.is_dtensor and self.mesh_info.is_spmd_mesh:  # False
      ...
  if self.is_dtensor:  # False
      ...
  return self._init_sharding_spec_plain(param, fsdp_placement)  # takes this path
  5. _init_sharding_spec_plain runs — shards the param on the DP submesh as a plain tensor, same as normal FSDP.

  So it silently works but degrades to regular FSDP on the DP submesh. The is_spmd_mesh flag is True on mesh_info, but every SPMD-specific code path in FSDPParam is gated on self.is_dtensor first, so none of them trigger.

  This is a silent misconfiguration — the user passed dp_mesh_dim_names expecting full SPMD mesh behavior, but their params aren't DTensors so they just get vanilla FSDP. There's no validation that says "if you provided
  dp_mesh_dim_names, parameters must be DTensors on that mesh." Could be worth flagging as another review comment — a warning or error in _init_sharding_spec when mesh_info.is_spmd_mesh is True but is_dtensor is False would catch this
  mistake early.
```

File: test/distributed/_composable/fsdp/test_fully_shard_dtensor.py [hunk 3]
```diff
@@ -0,0 +1,400 @@
...
+        fully_shard(
+            model,
+            mesh=mesh,
+            dp_mesh_dim_names=DataParallelMeshDims(shard="fsdp"),
+        )
+
+        replicate(
```
Comment (weifengpy) [inline test/distributed/_composable/fsdp/test_fully_shard_dtensor.py]: non-blocking but a plan question: I know this replicate() is for unit test purpose. what's the plan for replicate_with_fsdp.py to accept spmd mesh ? that's the replicate we use in titan

File: test/distributed/_composable/fsdp/test_fully_shard_dtensor.py [hunk 4]
```diff
@@ -0,0 +1,400 @@
...
+            ref_model,
+            device_ids=[self.rank] if device_type.type != "cpu" else None,
+            process_group=dp_pg,
+        )
+
+        self._run_train_parity(model, ref_model, dp_pg, mesh=mesh)
+
+    @skip_if_lt_x_gpu(4)
+    def test_hsdp_dtensor_train_parity(self):
+        """HSDP with 2D mesh, all params DTensors with (Replicate, Replicate)."""
```
Comment (weifengpy) [inline test/distributed/_composable/fsdp/test_fully_shard_dtensor.py]: also add tests for HSDP+TP on 3D SPMD mesh?

File: test/distributed/_composable/fsdp/test_fully_shard_dtensor.py [hunk 5]
```diff
@@ -0,0 +1,400 @@
...
+
+        torch.manual_seed(42)
+        model = MLP(16, device=device_type)
+        ref_model = copy.deepcopy(model)
+
+        distribute_module(model, mesh)
+        fully_shard(
+            model,
+            mesh=mesh,
+            dp_mesh_dim_names=DataParallelMeshDims(replicate="ddp"),
```
Comment (weifengpy) [inline test/distributed/_composable/fsdp/test_fully_shard_dtensor.py]: talked about this offline to support replicate=("ddp0", "ddp1")

File: test/distributed/_composable/fsdp/test_fully_shard_dtensor.py [hunk 6]
```diff
@@ -0,0 +1,400 @@
...
+        )
+
+        for param in model.parameters():
+            self.assertIsInstance(param, DTensor)
+            self.assertEqual(param.device_mesh, mesh)
+            self.assertEqual(len(param.placements), 1)
+            self.assertIsInstance(param.placements[0], Shard)
+
+    @skip_if_lt_x_gpu(2)
+    def test_validation_non_replicate_dp_placement(self):
```
Comment (weifengpy) [inline test/distributed/_composable/fsdp/test_fully_shard_dtensor.py:342]: what about putting Shard(0) on the "ddp" dim ?

File: torch/distributed/fsdp/_fully_shard/_fsdp_init.py [hunk 7]
```diff
@@ -61,18 +84,70 @@ def _validate_mesh(mesh: "DeviceMesh") -> None:
...
+        dp_mesh,
+        shard_mesh_dim=1,
+        replicate_mesh_dim=0,
+        dp_mesh_dim_names=dp_mesh_dim_names,
+        spmd_mesh=mesh,
+    )
+
+
 def _get_post_forward_mesh_info(
     reshard_after_forward: bool | int, mesh_info: FSDPMeshInfo
```
Comment (weifengpy) [inline torch/distributed/fsdp/_fully_shard/_fsdp_init.py:153]: from claude
```
reshard_after_forward=int (partial resharding): This is where the risk is. Look at lines 187-193:

  post_forward_mesh_tensor = mesh_info.mesh.mesh.view(-1, reshard_after_forward)
  post_forward_mesh = DeviceMesh(
      mesh_info.mesh.device_type, post_forward_mesh_tensor
  )
  post_forward_mesh_info = HSDPMeshInfo(
      post_forward_mesh, shard_mesh_dim=1, replicate_mesh_dim=0
  )

  In SPMD mode, mesh_info.mesh is the DP submesh (extracted from the full SPMD mesh), not the original mesh. This HSDPMeshInfo is constructed without dp_mesh_dim_names or spmd_mesh, so is_spmd_mesh would be False. Then in
  _init_sharded_post_forward_param_metadata (line 477), param._local_tensor is extracted assuming the TP/SPMD structure, but the post_forward_mesh_info has no awareness of the SPMD mesh. This is a plausible source of bugs — the
  post-forward resharding logic was written for the non-SPMD world and hasn't been audited for SPMD compatibility.

  So: the int case is a real gap. It's a feature specifically for large-scale training (partial resharding to reduce memory while keeping some communication overlap), which is exactly the scenario where SPMD meshes are most likely to be
   used. Worth calling out.
```

**Review by weifengpy — COMMENTED** [review]
done with the review - totally understand it's in draft yet. just want to have a complete review first

File: test/distributed/_composable/fsdp/test_fully_shard_dtensor.py (hunk 1 above)
Comment (fegin) [reply test/distributed/_composable/fsdp/test_fully_shard_dtensor.py:80]: Think about it is DDP, if no fully_shard is applied. `fully_shard` is just an optimization that perform the `allgather` and `reduce_scatter`. Without `fully_shard`, it will still work correctly due to DTensor.

File: test/distributed/_composable/fsdp/test_fully_shard_dtensor.py (hunk 3 above)
Comment (fegin) [reply test/distributed/_composable/fsdp/test_fully_shard_dtensor.py]: That's a good point. We should be able to do that as well. Let me see if I can include that in this PR.

File: torch/distributed/fsdp/_fully_shard/_fsdp_init.py (hunk 7 above)
Comment (fegin) [reply torch/distributed/fsdp/_fully_shard/_fsdp_init.py:153]: We don't support it if TP is enabled with the original code path. I don't think I'm going to extend this feature for DTensor in this PR. The internal branch also remove `reshard_after_forward=int` case afaik. I'll add an exception in the init phase.

File: test/distributed/_composable/fsdp/test_fully_shard_dtensor.py (hunk 2 above)
Comment (fegin) [reply test/distributed/_composable/fsdp/test_fully_shard_dtensor.py]: We should check and raise an explicit exception. Will add this.

File: torch/distributed/fsdp/_fully_shard/_fsdp_api.py [hunk 8]
```diff
@@ -126,6 +126,43 @@ def __call__(
...
+    """
+    Specifies which dimensions of a full SPMD :class:`DeviceMesh` correspond to
+    data parallelism when using :func:`fully_shard` whose parameters are already
+    DTensors on that mesh.
+
+    Attributes:
+        shard (Optional[Union[str, tuple[str, ...]]]): Mesh dimension name(s)
+            that FSDP shards parameters on. If a tuple of names, those dims
+            are flattened into a single shard dimension. At least one of
+            ``shard`` and ``replicate`` must be set.
```
Comment (wconstab) [inline torch/distributed/fsdp/_fully_shard/_fsdp_api.py:140]: add a __post__init__ to validate this?

File: test/distributed/_composable/fsdp/test_fully_shard_dtensor.py [hunk 9]
```diff
@@ -0,0 +1,645 @@
...
+
+        torch.manual_seed(42)
+        model = MLP(16, device=device_type)
+        ref_model = copy.deepcopy(model)
+
+        distribute_module(model, mesh)
+        fully_shard(
+            model,
+            mesh=mesh,
+            dp_mesh_dims=DataParallelMeshDims(shard=("dp0", "dp1")),
```
Comment (wconstab) [inline test/distributed/_composable/fsdp/test_fully_shard_dtensor.py]: I think it would be worth replacing most of these tests with one test using parametrize or using a loop over a list of test cases.  the test case can define the mesh names+shapes, basically.  Then there is much less LOC of boilerplate and easier to see in one place what cases have been covered.

File: torch/distributed/fsdp/_fully_shard/_fsdp_init.py [hunk 10]
```diff
@@ -61,18 +84,70 @@ def _validate_mesh(mesh: "DeviceMesh") -> None:
...
+    mesh: "DeviceMesh",
+    dp_mesh_dims: "DataParallelMeshDims",
+) -> "DataParallelMeshInfo":
+    shard_names = dp_mesh_dims.shard_names
+    replicate_names = dp_mesh_dims.replicate_names
+
+    def _get_submesh(names: tuple[str, ...]) -> "DeviceMesh":
+        if len(names) == 1:
+            return mesh[names[0]]
+        return mesh[names]._flatten("_".join(names))
```
Comment (wconstab) [inline torch/distributed/fsdp/_fully_shard/_fsdp_init.py:121]: does fsdp already have the convention of creating new flattened meshes (with PGs, which is what i assume this line does), or is this new behavior?

File: torch/distributed/fsdp/_fully_shard/_fsdp_init.py [hunk 11]
```diff
@@ -61,18 +84,70 @@ def _validate_mesh(mesh: "DeviceMesh") -> None:
...
+
+    if len(shard_names) == 0:  # DDP
+        if len(replicate_names) == 0:
```
Comment (wconstab) [inline torch/distributed/fsdp/_fully_shard/_fsdp_init.py]: we wouldn't need this if we validate earlier in __post_init_

File: torch/distributed/fsdp/_fully_shard/_fsdp_common.py [hunk 12]
```diff
@@ -31,12 +33,16 @@ class DataParallelMeshInfo:
     mesh: DeviceMesh
     shard_mesh_dim: int | None = None
     replicate_mesh_dim: int | None = None
+    dp_mesh_dims: DataParallelMeshDims | None = None
+    spmd_mesh: DeviceMesh | None = field(default=None, repr=False)
+    is_spmd_mesh: bool = field(default=False, init=False, repr=False)
```
Comment (wconstab) [inline torch/distributed/fsdp/_fully_shard/_fsdp_common.py:41]: i wonder if we could validate that the mesh is actually an spmd mesh. i guess not. we can not compare it to the world since we don't know if PP is involved and we can't assume the whole world is part of SPMD. it is important that users don't accidentally pass a submesh that excludes any other spmd dims here though right?

**Review by wconstab — COMMENTED** [review]
mostly lgtm. @weifengpy do you want to do the final review?

**Review by weifengpy — APPROVED** [review]
thanks for resolving my long list of comments!

File: torch/distributed/fsdp/_fully_shard/_fsdp_api.py (hunk 8 above)
Comment (fegin) [reply torch/distributed/fsdp/_fully_shard/_fsdp_api.py:140]: sgtm

File: torch/distributed/fsdp/_fully_shard/_fsdp_init.py (hunk 10 above)
Comment (fegin) [reply torch/distributed/fsdp/_fully_shard/_fsdp_init.py:121]: This is a new behavior. The existing behavior is that users need to "flatten" CP and FSDP by themselves.

File: torch/distributed/fsdp/_fully_shard/_fsdp_common.py (hunk 12 above)
Comment (fegin) [reply torch/distributed/fsdp/_fully_shard/_fsdp_common.py:41]: We could do some validations, but it is not possible to the accurate validation as you mentioned, if PP exists.
